src/rt/dmain2.d

```diff
@@ -320,21 +320,134 @@ extern (C) CArgs rt_cArgs() @nogc
+                args[i] = totalArgsBuff[j .. j + len];
+                if (len == 0)
+                    continue;
+                j += len;
+                assert(j <= totalArgsLength);
+                WideCharToMultiByte(CP_UTF8, 0, &wargs[i][0], cast(int) wlen, &args[i][0], len, null, null);
+            }
+        }
+        LocalFree(wargs);
+        wargs = null;
+        wargc = 0;
+    }
+    else version (Posix)
+    {
+        // Allocate args[] on the stack
+        char[][] args = (cast(char[]*) alloca(argc * (char[]).sizeof))[0 .. argc];
+
+        size_t totalArgsLength = 0;
+        foreach (i, ref arg; args)
+        {
+            arg = argv[i][0 .. strlen(argv[i])];
+            totalArgsLength += arg.length;
+        }
+    }
+    else
+        static assert(0);
+
+    return _d_run_main2(args, totalArgsLength, mainFunc);
+}
+
+/**
+ * Windows-specific version for wide command-line arguments, e.g.,
+ * from a wmain/wWinMain C entry point.
+ * This wide version uses the specified arguments, unlike narrow
+ * _d_run_main which uses the actual (wide) process arguments instead.
+ */
+version (Windows)
+extern (C) int _d_wrun_main(int argc, wchar** wargv, MainFunc mainFunc)
+{
+     // Allocate args[] on the stack
+    char[][] args = (cast(char[]*) alloca(argc * (char[]).sizeof))[0 .. argc];
+
+    // 1st pass: compute each argument's length as UTF-16 and UTF-8
+    size_t totalArgsLength = 0;
+    foreach (i; 0 .. argc)
+    {
+        const warg = wargv[i];
+        const size_t wlen = wcslen(warg) + 1; // incl. terminating null
+        assert(wlen <= cast(size_t) int.max, "wlen cannot exceed int.max");
+        const int len = WideCharToMultiByte(CP_UTF8, 0, warg, cast(int) wlen, null, 0, null, null);
+        args[i] = (cast(char*) wlen)[0 .. len]; // args[i].ptr = wlen, args[i].length = len
+        totalArgsLength += len;
+    }
+
+    // Allocate a single buffer for all (null-terminated) argument strings in UTF-8 on the stack
+    char* utf8Buffer = cast(char*) alloca(totalArgsLength);
+
+    // 2nd pass: convert to UTF-8 and finalize `args`
+    char* utf8 = utf8Buffer;
+    foreach (i; 0 .. argc)
+    {
+        const wlen = cast(int) args[i].ptr;
+        const len = cast(int) args[i].length;
+        WideCharToMultiByte(CP_UTF8, 0, wargv[i], wlen, utf8, len, null, null);
+        args[i] = utf8[0 .. len-1]; // excl. terminating null
+        utf8 += len;
+    }
+
+    // Set C argc/argv; argv is a new stack-allocated array of UTF-8 C strings
+    char*[] argv = (cast(char**) alloca(argc * (char*).sizeof))[0 .. argc];
+    foreach (i, ref arg; argv)
+        arg = args[i].ptr;
+    _cArgs.argc = argc;
+    _cArgs.argv = argv.ptr;
+
+    totalArgsLength -= argc; // excl. null terminator per arg
+    return _d_run_main2(args, totalArgsLength, mainFunc);
+}
+
+private extern (C) int _d_run_main2(char[][] args, size_t totalArgsLength, MainFunc mainFunc)
```

Comment:
passing `totalArgsLength` here looks a bit weird, but so is the additional copy of the arguments below. Can we move the latter up to the calling functions and simplify something from that?

Reply:
Not related to this PR, but I dislike the additional copy too (IMO just a signature problem - calling _Dmain with `string[] args` wouldn't need it).

Reply:
> IMO just a signature problem

Not a signature issue only, but the guarantee that the arguments are actually immutable. Without the copy, cArgs is a mutable alias to the same memory.

Reply:
`cArgs.argv` could be adapted too (`immutable(char)**`), or `string[] args` be relaxed to `const(char)[]`.

Reply:
> cArgs is a mutable alias to the same memory

It's actually a mutable alias to `string[] rt_args()` (only exception being Windows with narrow entry point), so that problem (potential breakage of immutability) already exists. `rt_args()` resolves to the first `char[][] args`, and so also includes the `--DRT-*` options, which isn't clear by its documentation (`To get out-of-band access to the args[] passed to main()`).